prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+	attr_ex.pd = domain->pd;
+	attr_ex.xrcd = domain->xrc.xrcd;
+	if (rdma_create_qp_ex(ep->tgt_id, &attr_ex)) {
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Physical XRC TGT QP, rdma_create_ep_ex()",
+				 errno);
+		return -errno;
+	}
+	ep->tgt_ibv_qp = ep->tgt_id->qp;
+
+	return FI_SUCCESS;
+}
+
+static int fi_ibv_put_tgt_qp(struct fi_ibv_ep *ep)
+{
+	int ret;
+
+	if (!ep->tgt_ibv_qp)
+		return FI_SUCCESS;
+
+	/* The kernel will not destroy the detached TGT QP until all
+	 * shared opens have called ibv_destroy_qp. */
+	ret = ibv_destroy_qp(ep->tgt_ibv_qp);
+	if (ret) {
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Close XRC TGT QP, ibv_destroy_qp()", errno);
+		return -errno;
+	}
+	ep->tgt_ibv_qp = NULL;
+
+	return FI_SUCCESS;
+}
+
+int fi_ibv_ep_destroy_xrc_qp(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+
+	if (ep->ibv_qp) {
+		fastlock_acquire(&domain->xrc.ini_mgmt_lock);
+		fi_ibv_put_shared_ini_conn(ep);
+		fastlock_release(&domain->xrc.ini_mgmt_lock);
+	}
+	if (ep->id) {
+		rdma_destroy_id(ep->id);
+		ep->id = NULL;
+	}
+	if (ep->tgt_ibv_qp)
+		fi_ibv_put_tgt_qp(ep);
+
+	if (ep->tgt_id) {
+		rdma_destroy_id(ep->tgt_id);
+		ep->tgt_id = NULL;
+	}
+	return 0;
+}
+
+static int fi_ibv_ini_conn_compare(struct ofi_rbmap *map, void *key, void *data)
+{
+	struct fi_ibv_ini_shared_conn *ini_conn = data;
+	struct fi_ibv_ini_conn_key *_key = key;
+	int ret;
+
+	assert(_key->addr->sa_family == ini_conn->peer_addr->sa_family);
+
+	/* Only interested in the interface address and TX CQ */
+	switch (_key->addr->sa_family) {
+	case AF_INET:
+		ret = memcmp(&ofi_sin_addr(_key->addr),
+			     &ofi_sin_addr(ini_conn->peer_addr),
+			     sizeof(ofi_sin_addr(_key->addr)));
+		break;
+	case AF_INET6:
+		ret = memcmp(&ofi_sin6_addr(_key->addr),
+			     &ofi_sin6_addr(ini_conn->peer_addr),
+			     sizeof(ofi_sin6_addr(_key->addr)));
+		break;
+	default:
+		VERBS_WARN(FI_LOG_FABRIC, "Unsupported address format\n");
+		assert(0);
+		return -FI_EINVAL;
+	}
+	if (ret)
+		return ret;
+
+	return _key->tx_cq < ini_conn->tx_cq ?
+			-1 : _key->tx_cq > ini_conn->tx_cq;
+}
+
+int fi_ibv_domain_xrc_init(struct fi_ibv_domain *domain)
+{
+	struct ibv_xrcd_init_attr attr;
+	int ret;
+
+	domain->xrc.xrcd_fd = -1;
+	if (fi_ibv_gl_data.msg.xrcd_filename) {
+		domain->xrc.xrcd_fd = open(fi_ibv_gl_data.msg.xrcd_filename,
+				       O_CREAT, S_IWUSR | S_IRUSR);
+		if (domain->xrc.xrcd_fd < 0) {
+			VERBS_INFO_ERRNO(FI_LOG_DOMAIN,
+					 "XRCD file open", errno);
+			return -errno;
+		}
+	}
+
+	attr.comp_mask = IBV_XRCD_INIT_ATTR_FD | IBV_XRCD_INIT_ATTR_OFLAGS;
+	attr.fd = domain->xrc.xrcd_fd;
+	attr.oflags = O_CREAT;
+	domain->xrc.xrcd = ibv_open_xrcd(domain->verbs, &attr);
+	if (!domain->xrc.xrcd) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_DOMAIN, "ibv_open_xrcd", errno);
+		goto xrcd_err;
+	}
+
+	fastlock_init(&domain->xrc.ini_mgmt_lock);
+	domain->xrc.ini_conn_rbmap = calloc(1,
+			sizeof(*domain->xrc.ini_conn_rbmap));
```

Comment:
Why not just allocated as part of larger structure?

Reply:
The existing OFI RBTree cleanup function, ofi_rbmap_cleanup(), assumes the RB map was dynamically allocated and frees the map after it deletes any nodes. I could change the XRC cleanup code to just delete the RB Tree, but that does require knowledge of the RBTree structure.